WinImageBuilder.psm1

```diff
@@ -157,177 +203,245 @@ function CreateBCDBootConfig($systemDrive, $windowsDrive, $diskLayout, $image)
+            Copy-Item -Recurse $src $dst
+        } else {
+            throw "The Windows curtin hooks module is not present.
+                Please run git submodule update --init " }
     }
 }
 
-function DownloadCloudbaseInit($resourcesDir, $osArch)
-{
-    Write-Output "Downloading Cloudbase-Init..."
+function Download-CloudbaseInit {
+    Param(
+        [Parameter(Mandatory=$true)]
+        [string]$resourcesDir,
+        [Parameter(Mandatory=$true)]
+        [string]$osArch
+    )
+    Write-Host "Downloading Cloudbase-Init..."
 
-    if($osArch -eq "AMD64")
-    {
+    if ($osArch -eq "AMD64") {
         $CloudbaseInitMsi = "CloudbaseInitSetup_Stable_x64.msi"
-    }
-    else
-    {
+    } else {
         $CloudbaseInitMsi = "CloudbaseInitSetup_Stable_x86.msi"
     }
 
     $CloudbaseInitMsiPath = "$resourcesDir\CloudbaseInit.msi"
     $CloudbaseInitMsiUrl = "https://www.cloudbase.it/downloads/$CloudbaseInitMsi"
 
-    ExecRetry {
+    Execute-Retry {
         (New-Object System.Net.WebClient).DownloadFile($CloudbaseInitMsiUrl, $CloudbaseInitMsiPath)
     }
-
 }
 
-function GenerateConfigFile($resourcesDir, $installUpdates)
-{
+function Generate-ConfigFile {
+    Param(
+        [Parameter(Mandatory=$true)]
+        [string]$resourcesDir,
+        [Parameter(Mandatory=$true)]
+        [hashtable]$values
+    )
+
     $configIniPath = "$resourcesDir\config.ini"
     Import-Module "$localResourcesDir\ini.psm1"
-    Set-IniFileValue -Path $configIniPath -Section "DEFAULT" -Key "InstallUpdates" -Value $installUpdates
+    foreach ($i in $values.GetEnumerator()) {
+        Set-IniFileValue -Path $configIniPath -Section "DEFAULT" -Key $i.Key -Value $i.Value
+    }
 }
 
-function AddDriversToImage($winImagePath, $driversPath)
-{
+function Add-DriversToImage {
+    Param(
+        [Parameter(Mandatory=$true)]
+        [string]$winImagePath,
+        [Parameter(Mandatory=$true)]
+        [string]$driversPath
+    )
     Write-Output ('Adding drivers from "{0}" to image "{1}"' -f $driversPath, $winImagePath)
-    Add-WindowsDriver -Path $winImagePath -Driver $driversPath -ForceUnsigned -Recurse
-    #& Dism.exe /image:${winImagePath} /Add-Driver /driver:${driversPath} /ForceUnsigned /recurse
-    #if ($LASTEXITCODE) { throw "Dism failed to add drivers from: $driversPath" }
-}
-
-function SetProductKeyInImage($winImagePath, $productKey)
-{
-    Set-WindowsProductKey -Path $winImagePath -ProductKey $productKey
+    & Dism.exe /image:${winImagePath} /Add-Driver /driver:${driversPath} /ForceUnsigned /recurse
+    if ($LASTEXITCODE) {
+        throw "Dism failed to add drivers from: $driversPath"
+    }
 }
 
-function EnableFeaturesInImage($winImagePath, $featureNames)
-{
-    if($featureNames)
-    {
+function Enable-FeaturesInImage {
+    Param(
+        [Parameter(Mandatory=$true)]
+        [string]$winImagePath,
+        [Parameter(Mandatory=$true)]
+        [array]$featureNames
+    )
+    if ($featureNames) {

```

Comment:
I think this would look better like:

```powershell
$cmd = @(
    "Dism.exe",
    ("/image:{0}" -f ${winImagePath}),
    "/Enable-Feature"
)
foreach ($featureName in $featureNames) {
    $cmd += ("/FeatureName:{0}" -f $featureName)
}
& $cmd[0] $cmd[1..$cmd.Length]
if ($LASTEXITCODE) { throw "Dism failed to enable features: $featureNames" }
```